### PR TITLE
Simplify logging using static invocations and bytecode transformation

### LINKED ISSUE #10929: Logging with Panache
ATM this is how everyone must set up logging:

```java
class X {
    private static final Logger log = Logger.getLogger(X.class);
    // ...
    void f(){
       log.info("foo");
    }
}
```

Which has the same problem as CDI in terms of requiring you to get out of your method, go up to find where to define statics, remember the magic incantation, and then figure out where you wanted to call it. It ruins the movie.

But unlike CDI injection, logging is something that happens super extra often. This is why a lot of people go for `System.err.println` rather than logging during dev, because it's so much easier.

Let's fix this.

I'd like to make it so this:

```java
class X {
    // ...
    void f(){
       QuarkusLogger.info("foo");
    }
}
```

is turned into the previous example by bytecode transformation. We should be able to optimise this like we optimise other transformations by checking if the class has a quarkus marker and has `QuarkusLogger` in its constant pool.

I hope this ease of use will lead to more people leaving logging info in their code after debugging, resulting in better logging, as opposed to all `println` being just removed after debugging, never replaced with proper logging.

WDYT?

### PR DESCRIPTION
**Pull request by Ladicek**
Resolves #10929.

### CODE REVIEW

**Comment by Ladicek** [comment]
Creating as a draft to let CI run in my fork, and perhaps to solicit some early reviews. The entire thing works, but need to decide where to put the code (probably shouldn't be a standalone extension, but part of `core`?) and documentation (probably in https://quarkus.io/guides/logging).

**Comment by FroMage** [comment]
VERY NICE!!!

File: extensions/logging/deployment/src/main/java/io/quarkus/logging/deployment/QuarkusLogProcessor.java
Comment (geoand) [inline extensions/logging/deployment/src/main/java/io/quarkus/logging/deployment/QuarkusLogProcessor.java:40]: Can we get some Javadoc explaining what the visitor does?

File: independent-projects/resteasy-reactive/pom.xml
Comment (geoand) [inline independent-projects/resteasy-reactive/pom.xml:39]: I sure hope this goes away :)

**Comment by geoand** [comment]
Very nice indeed!

File: extensions/logging/deployment/src/main/java/io/quarkus/logging/deployment/QuarkusLogProcessor.java
Comment (Ladicek) [reply extensions/logging/deployment/src/main/java/io/quarkus/logging/deployment/QuarkusLogProcessor.java:40]: I would have hoped that the name is enough, but I'll add something :-)

File: independent-projects/resteasy-reactive/pom.xml
Comment (Ladicek) [reply independent-projects/resteasy-reactive/pom.xml:39]: Jason says here https://github.com/wildfly/jandex/issues/84#issuecomment-765692608 that Jandex 3.0 will include this facility. (I should have mentioned this above, it's one more reason why this is a draft. When Jandex is released, this will obviously change, and the srcdeps stuff will go away.)

**Comment by Ladicek** [comment]
One thing I wanted to ask everyone: should this be an extra extension (it is right now), or part of something else? There's just one class in `runtime`, so perhaps this could move to `core`, so that it's always available? There's a ton of other stuff in `core/runtime` in `io.quarkus.runtime.logging`, so not exactly sure that's the best place either...

**Comment by geoand** [comment]
> One thing I wanted to ask everyone: should this be an extra extension (it is right now), or part of something else?

It feels like a core thing to me

**Comment by FroMage** [comment]
Definitely core material IMO.

**Comment by FroMage** [comment]
We want people to have this OOTB and it doesn't cost if people don't use it and there are no new dependencies.

**Comment by Ladicek** [comment]
Agree, thanks, I'll move it around.

**Comment by Ladicek** [comment]
Moved to core and added documentation.

**Comment by Ladicek** [comment]
And also added a devmode test.

**Comment by FroMage** [comment]
So why is this still a draft?

**Comment by Ladicek** [comment]
Waiting for @n1hility to merge https://github.com/wildfly/jandex/pull/112 and release Jandex 3.0 :-)

**Comment by FroMage** [comment]
Mmm, who else can merge this?

**Comment by Sanne** [comment]
Reviving

**Comment by Ladicek** [comment]
I see the PR to Jandex was merged, but no Jandex release yet, so this will still have to wait a bit :-)

**Comment by FroMage** [comment]
Yay, let's revive this and get it merged!

**Comment by Ladicek** [comment]
I've rebased my branch, finalized the code and updated this PR. I'll mark it ready as soon as Jandex 2.4.0.Final reaches Central.

File: integration-tests/pom.xml
Comment (gastaldi) [inline integration-tests/pom.xml:275]: You need to add this new module to an existing slot in https://github.com/quarkusio/quarkus/blob/main/.github/native-tests.json

Comment (gastaldi) [reply integration-tests/pom.xml:275]: Ah there are no native tests, maybe that's not necessary

File: integration-tests/logging-panache/src/test/java/io/quarkus/logging/GenerateAllLogUsages.java
Comment (gastaldi) [inline integration-tests/logging-panache/src/test/java/io/quarkus/logging/GenerateAllLogUsages.java:11]: Why this is not a test class?

File: docs/src/main/asciidoc/logging.adoc
Comment (gastaldi) [inline docs/src/main/asciidoc/logging.adoc:80]: The title might need to change now that this feature lives in core?

**Review by gastaldi — APPROVED** [review]
Works as advertised! Good job!

**Comment by gastaldi** [comment]
Let's merge this once https://github.com/quarkusio/quarkus/pull/14586#discussion_r698809162 is fixed

**Review by gastaldi — CHANGES_REQUESTED** [review]
Waiting for the documentation update. Feel free to dismiss this request for changes and merge once it's done

File: core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingWithPanacheProcessor.java
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingWithPanacheProcessor.java:19]: This class definitely needs to be renamed

File: integration-tests/logging-panache/src/test/java/io/quarkus/logging/GenerateAllLogUsages.java
Comment (Ladicek) [reply integration-tests/logging-panache/src/test/java/io/quarkus/logging/GenerateAllLogUsages.java:11]: Because it's not? It generates a bean with all possible `Log.*` invocations, which is then used in a test.

File: docs/src/main/asciidoc/logging.adoc
Comment (Ladicek) [reply docs/src/main/asciidoc/logging.adoc:80]: I don't really care, but I also don't really know. My understanding is that the "Panache" term (as much as I dislike it) is the Quarkus "brand" for "hey look how Quarkus can simplify your life with a little bit of magic", in which case, "Logging with Panache" seems like a perfect fit?

Comment (gastaldi) [reply docs/src/main/asciidoc/logging.adoc:80]: Now I am confused, isn't Panache related to ORM specifically?

Comment (Ladicek) [reply docs/src/main/asciidoc/logging.adoc:80]: They way I understand it, this is a widespread misconception :-) @FroMage any comments on that? Thanks!

Comment (gsmet) [reply docs/src/main/asciidoc/logging.adoc:80]: Yeah @Ladicek is right. It's supposed to be a general brand for our opinionated magic.

Comment (gastaldi) [reply docs/src/main/asciidoc/logging.adoc:80]: I'm afraid our users may be confused with this as well, because we don´t have anything (at least in our documentation) about Panache that is not related to ORM.

Comment (gsmet) [reply docs/src/main/asciidoc/logging.adoc:80]: Well, we have MongoDB with Panache.

Comment (gastaldi) [reply docs/src/main/asciidoc/logging.adoc:80]: > Well, we have MongoDB with Panache.

Still ORM no? :) Anyway, I am fine with it as long as the message is not confusing to our users

Comment (Ladicek) [reply docs/src/main/asciidoc/logging.adoc:80]: I'm actually with @gastaldi on this, it is confusing, plus I never really liked the term :rofl:, so... as I said, I don't really care, but I also don't really know.

**Comment by Ladicek** [comment]
Jandex Maven plugin 1.2.0 released, I'll bump it in this PR when it reaches Central.

File: docs/src/main/asciidoc/logging.adoc
Comment (FroMage) [reply docs/src/main/asciidoc/logging.adoc:80]: It's supposed to be generic and related to magic, so not related to ORM, which is why we're always careful to not abbreviate `ORM with Panache` to just `Panache`.

Comment (FroMage) [inline docs/src/main/asciidoc/logging.adoc:80]: If this is the simpler option, perhaps we should move it up and recommend it against the traditional injection?

**Review by FroMage — APPROVED** [review]
Looks great, congrats!

File: docs/src/main/asciidoc/logging.adoc
Comment (Ladicek) [reply docs/src/main/asciidoc/logging.adoc:80]: Good point, I'll move it up.

Comment (Ladicek) [reply docs/src/main/asciidoc/logging.adoc]: Done.

**Review by gastaldi — APPROVED** [review]
Let's get this merged after the plugin version bump. Good job @Ladicek !

**Comment by Ladicek** [comment]
Jandex Maven plugin updated to 1.2.0, yay!

**Pull request merged**
